### PR TITLE
[DO NOT MERGE] OV20 PrePostProcessor draft

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: ngraph/core/src/pre_post_process/pre_post_process.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            auto new_param_type = param->get_element_type();
+            auto new_param_shape = param->get_partial_shape();
+            auto new_param_layout = PartialLayout();
+            if (in_context.m_tensor_data.m_type != ov::element::dynamic) {
+                new_param_type = in_context.m_tensor_data.m_type;
+            } else {
+                // Look for pre_process data
+                for (auto action : in_context.m_preprocess_data.m_actions) {
+                    if (std::dynamic_pointer_cast<PreProcessConvertElementType>(action)) {
+                        throw ngraph::ngraph_error("Can't convert element type from unknown type");
```
Comment (ilya-lavrenov) [inline ngraph/core/src/pre_post_process/pre_post_process.cpp:134]: here we have an assumption that no other preprocessing operations convert element types. It's true for now I believe, but in general back propagation can help with it

File: ngraph/test/pre_post_process.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    auto f = create_simple_function(ngraph::element::f32, ngraph::Shape{1, 3, 4, 5});
+    ngraph::pass::VisualizeTree("before.png").run_on_function(f);  // Visualize the nGraph function to an image
+    f = PrePostProcessor(f)
+            .in()
+            .preprocess()
+            .scale(2.0f)
+            .scale(4.0f)
+            .network()
+            .set_layout(PartialLayout("NCHW"))
+            .build();
```
Comment (ilya-lavrenov) [inline ngraph/test/pre_post_process.cpp:34]: can we create a reference ngraph function and compare with it? 
can we also use different indentation for different sections? clang-format can be turned off such these places

File: ngraph/test/pre_post_process.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    ASSERT_EQ(f->get_output_element_type(0), element::i8);
+    ngraph::pass::VisualizeTree("before2.png").run_on_function(f);  // Visualize the nGraph function to an image
+    f = PrePostProcessor(f)
+            .in()
+            .tensor()
+            .set_element_type(element::i16)
+            .preprocess()
+            .convert_element_type(element::f32)
+            .scale(0.2f)
+            .network()
```
Comment (ilya-lavrenov) [inline ngraph/test/pre_post_process.cpp:53]: network can be omitted here

File: ngraph/test/pre_post_process.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+TEST(pre_post_process, scale_vector_no_channels_layout) {
+    auto f = create_simple_function(element::f32, ngraph::PartialShape{Dimension::dynamic(), 4, 4, 3});
+    ASSERT_EQ(f->get_output_element_type(0), element::f32);
+    ASSERT_ANY_THROW(f = PrePostProcessor(f)
+                             .in()
+                             .tensor()
+                             .set_layout(PartialLayout("NHW?"))
+                             .preprocess()
+                             .scale({0.1f, 0.2f, 0.3f})
+                             .build();
```
Comment (ilya-lavrenov) [inline ngraph/test/pre_post_process.cpp:135]: can we add test that w/o layout scale throws exception?

File: ngraph/core/src/pre_post_process/pre_post_process.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+OutProcessor PrePostProcessorBase::out() {
+    return OutProcessor(std::move(m_impl));
+}
+
+std::shared_ptr<Function> PrePostProcessorBase::build() {
+    // TODO: maybe wrap this logic with 'FunctionPass'???
+    for (auto in_context : m_impl->in_contexts) {
+        if (in_context.is_default()) {
+            auto param = m_impl->m_function->get_parameters().front();
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/core/src/pre_post_process/pre_post_process.cpp:122]: It is very rough prototype, yes this check is needed to get appropriate parameter. Other logic is common for all 'in_contexts'

File: ngraph/core/src/pre_post_process/pre_post_process.cpp (hunk 1 above)
Comment (nosovmik) [reply ngraph/core/src/pre_post_process/pre_post_process.cpp:134]: Probably we should not raise an exception in this case and just leave 'user_tensor_element_type' as 'dynamic'. But can you please provide an example of back propagation?
Suppose we have function with 'i8' input parameter
`PrePostProcessor(f).in().preprocess().convert_element_type(f32).<doOther>().build()' - how we can guess source tensor element type using back propagation?

File: ngraph/test/pre_post_process.cpp (hunk 2 above)
Comment (nosovmik) [reply ngraph/test/pre_post_process.cpp:34]: yes, sure, currently it is very rough sketch

File: ngraph/test/pre_post_process.cpp (hunk 4 above)
Comment (nosovmik) [reply ngraph/test/pre_post_process.cpp:135]: yes, will do

**Pull request closed without merging**